common/djangoapps/student/models.py

```diff
@@ -1700,69 +1700,39 @@ def upgrade_deadline(self):
         Returns:
             datetime|None
         """
-        log.debug('Schedules: Determining upgrade deadline for CourseEnrollment %d...', self.id)
-        if not CourseMode.is_mode_upgradeable(self.mode):
-            log.debug(
-                'Schedules: %s mode of %s is not upgradeable. Returning None for upgrade deadline.',
-                self.mode, self.course_id
-            )
-            return None
-
-        if self.dynamic_upgrade_deadline is not None:
-            return self.dynamic_upgrade_deadline
-
-        return self.course_upgrade_deadline
+        if CourseMode.is_mode_upgradeable(self.mode):
+            if self.dynamic_upgrade_deadline is not None:
+                return self.dynamic_upgrade_deadline
+            else:
+                return self.course_upgrade_deadline
```

Comment:
I actually tend to use the style that you're moving away from (w/ early returns). I'm ok w/ switching, but I do think that it's better to explicitly `return None` at the end of a function that is otherwise returning things, so that it's clear that you are deliberately returning `None`, and not just accidentally falling through all of the if statements.

Reply:
I guess it's a subjective preference.  For me, since I know Python defaults to `return None`, I prefer not making it explicit - and keeping the reader's focus on the non-empty path.
Having said that, I'm Ok with adding the empty returns at the end of explicitness.  I'll go ahead and make that change.  However, I do feel the rest of the revised code here is an easier flow to understand.  Curious to see what others think.  I'm happy to not merge this if these changes don't appeal to others.

Reply:
I think that the block is here simple enough where it doesn't matter, but in general I also prefer early returns and no nested ifs.

I also agree with returning an explicit `None` whenever the caller is expected to actually use the return value (as opposed to a function you call just for the side-effect).

Reply:
To elaborate further, I think that early-return can be confusing to follow if the conditionals are a long grab bag list of random things. But I think is especially useful when there's a lot of logic involved and you can break it up into various layers like:

```python
# Course policy
# (code here)

# User targeting
# (code here)

# Date conditions
# (code here)
```

But again, this one is short enough where I don't have a strong opinion either way.

Reply:
I do prefer the pattern of early returns when a function is long - as opposed to multiple nested if-else clauses - that you have to follow all the way down to see if any of the clauses change the return value.

However, if you look carefully, changing from early-returns to single-returns was not the objective of my change.  Rather, it was to make the logic of the functions easier to understand at a quick glance.  This is particularly true for `upgrade_deadline` and `dynamic_upgrade_deadline` -> by refactoring them into smaller functions and removing no longer needed log statements that now get in the way.

Yes, it's true that I may have gotten carried away with changing `verified_upgrade_link_is_valid`.  Since it's a small enough function, I didn't think it needed an early-return pattern - so I changed it to give focus to the logic.  Sure, this could be done either way.

Reply:
I actually have found those log statements very useful for debugging failing tests. I would advocate for keeping them if possible.